src/Controller/Admin/Improve/Design/LinkWidgetController.php

```diff
@@ -0,0 +1,252 @@
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse|array
+     * @throws \Exception
+     */
+    public function createProcessAction(Request $request)
+    {
+        return $this->processForm($request, 'Successful creation.');
+    }
+
+    /**
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="Access denied.")
+     * @Template("@Modules/ps_linklist/views/templates/admin/link_block/form.html.twig")
+     *
+     * @param Request $request
+     * @param $linkBlockId
+     *
+     * @return RedirectResponse|array
+     * @throws \Exception
+     */
+    public function editProcessAction(Request $request, $linkBlockId)
+    {
+        return $this->processForm($request, 'Successful update.', $linkBlockId);
+    }
+
+    /**
+     * @param int $linkBlockId
+     * @return RedirectResponse
+     */
+    public function deleteAction($linkBlockId)
+    {
+        $repository = $this->get('prestashop.module.link_block.repository');
+        $errors = [];
+        try {
+            $repository->delete($linkBlockId);
+        } catch (PrestaShopDatabaseException $e) {
+            $errors[] = [
+                'key' => "Could not delete #%i",
+                'domain' => 'Admin.Catalog.Notification',
+                'parameters' => [$linkBlockId],
+            ];
+        }
+
+        if (0 === count($errors)) {
+            $this->clearModuleCache();
+            $this->addFlash('success', $this->trans('Successful deletion.', 'Admin.Notifications.Success'));
+        } else {
+            $this->flashErrors($errors);
+        }
+
+        return $this->redirectToRoute('admin_link_widget_list');
+    }
+
+    /**
+     * @param Request  $request
+     * @param string   $successMessage
+     * @param int|null $linkBlockId
+     * @return array|RedirectResponse
+     * @throws \Exception
+     */
+    private function processForm(Request $request, $successMessage, $linkBlockId = null)
+    {
+        /** @var LinkBlockFormDataProvider $formProvider */
+        $formProvider = $this->get('prestashop.module.link_block.form_provider');
+        $formProvider->setIdLinkBlock($linkBlockId);
+
+        /** @var FormHandlerInterface $formHandler */
+        $formHandler = $this->get('prestashop.module.link_block.form_handler');
+        $form = $formHandler->getForm();
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted()) {
+            $data = $form->getData();
+            $saveErrors = $formHandler->save($data);
+
+            if (0 === count($saveErrors)) {
+                $this->addFlash('success', $this->trans($successMessage, 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_link_block_edit', ['linkBlockId' => $formProvider->getIdLinkBlock()]);
+            }
+
+            $this->flashErrors($saveErrors);
+        }
+
+        return [
+            'linkBlockForm' => $form->createView(),
```

Comment:
You keep the `$form->createView()` what is the interest to have another view to GET the form?

Reply:
I'm not sure I get what you mean I keep the form displaying in createProcess/editProcess to not lose the previously submitted data
And I could factorize the four endpoints create[GET,POST] and edit[GET,POST] but I find it cleaner to keep each end points separeted, yet I factorized the `private function processForm` since in this case they are managed the same way This might change when the new `EntityForm` will be available

Which makes me think that I have no hook for creation/update But there were none in the previous module Maybe this can wait the refacto where the hooks will be managed "natively" in the core